swarm_copy_tests/app/test_dependencies.py

```diff
@@ -0,0 +1,171 @@
+):
+    token = "fake_token"
+    file_name = "fake_file"
+    client = AsyncClient()
+
+    file_url = "https://fake_file_url"
+
+    monkeypatch.setenv(
+        "NEUROAGENT_KNOWLEDGE_GRAPH__HIERARCHY_URL", "http://fake_hierarchy_url.com"
+    )
+
+    settings = Settings(
+        knowledge_graph={"br_saving_path": tmp_path / "test_brain_region.json"}
+    )
+
+    json_response_url = {
+        "head": {"vars": ["file_url"]},
+        "results": {"bindings": [{"file_url": {"type": "uri", "value": file_url}}]},
+    }
+    with open(
+        Path(__file__).parent.parent.parent
+        / "tests"
+        / "data"
+        / "KG_brain_regions_hierarchy_test.json"
+    ) as fh:
+        json_response_file = json.load(fh)
+
+    httpx_mock.add_response(
+        url=settings.knowledge_graph.sparql_url, json=json_response_url
+    )
+    httpx_mock.add_response(url=file_url, json=json_response_file)
+
+    await get_update_kg_hierarchy(
+        token,
+        client,
+        settings,
+        file_name,
+    )
+
+    assert os.path.exists(settings.knowledge_graph.br_saving_path)
+
+
+@pytest.mark.asyncio
+async def test_get_cell_types_kg_hierarchy(
+    tmp_path, httpx_mock, monkeypatch, patch_required_env
+):
+    token = "fake_token"
+    file_name = "fake_file"
+    client = AsyncClient()
+
+    file_url = "https://fake_file_url"
+    monkeypatch.setenv(
+        "NEUROAGENT_KNOWLEDGE_GRAPH__HIERARCHY_URL", "http://fake_hierarchy_url.com"
+    )
+
+    settings = Settings(
+        knowledge_graph={"ct_saving_path": tmp_path / "test_cell_types_region.json"}
+    )
+
+    json_response_url = {
+        "head": {"vars": ["file_url"]},
+        "results": {"bindings": [{"file_url": {"type": "uri", "value": file_url}}]},
+    }
+    with open(
+        Path(__file__).parent.parent.parent
+        / "tests"
+        / "data"
+        / "kg_cell_types_hierarchy_test.json"
+    ) as fh:
+        json_response_file = json.load(fh)
+
+    httpx_mock.add_response(
+        url=settings.knowledge_graph.sparql_url, json=json_response_url
+    )
+    httpx_mock.add_response(url=file_url, json=json_response_file)
+
+    await get_cell_types_kg_hierarchy(
+        token,
+        client,
+        settings,
+        file_name,
+    )
+
+    assert os.path.exists(settings.knowledge_graph.ct_saving_path)
+
+
+def test_get_connection_string_full(monkeypatch, patch_required_env):
+    monkeypatch.setenv("NEUROAGENT_DB__PREFIX", "http://")
+    monkeypatch.setenv("NEUROAGENT_DB__USER", "John")
+    monkeypatch.setenv("NEUROAGENT_DB__PASSWORD", "Doe")
+    monkeypatch.setenv("NEUROAGENT_DB__HOST", "localhost")
+    monkeypatch.setenv("NEUROAGENT_DB__PORT", "5000")
+    monkeypatch.setenv("NEUROAGENT_DB__NAME", "test")
+
+    settings = Settings()
+    result = get_connection_string(settings)
+    assert (
+        result == "http://John:Doe@localhost:5000/test"
+    ), "must return fully formed connection string"
```

Comment:
Couple of tests are missing:
- `get_vlab_and_project`
- `get_starting_agent`
- `get_context_variables` (arguably no need to test it)
- `get_openai_client` (arguably no need to test it)
- `get_agents_routine` (arguably no need to test it)
- `get_kg_token`

Reply:
I added tests for 

- get_vlab_and_project
- get_starting_agent
- get_kg_token

For the other 3, I don't think there is much point.